src/librustc/ty/maps/plumbing.rs

```diff
@@ -564,3 +592,254 @@ macro_rules! define_provider_struct {
         }
     };
 }
+
+pub fn force_from_dep_node<'a, 'gcx, 'lcx>(tcx: TyCtxt<'a, 'gcx, 'lcx>,
+                                           dep_node: &DepNode)
+                                           -> bool {
+    use ty::maps::keys::Key;
+    use hir::def_id::LOCAL_CRATE;
+
+    // We should never get into the situation of having to force this from the DepNode.
+    // Since we cannot reconstruct the query key, we would always end up having to evaluate
+    // the first caller of this query that *is* reconstructible. This might very well be
+    // compile_codegen_unit() in which case we'd lose all re-use.
+    debug_assert!(dep_node.kind != DepKind::CodegenUnit,
+                  "calling force_from_dep_node() on DepKind::CodegenUnit");
+
+    if !dep_node.kind.can_reconstruct_query_key() {
+        return false
+    }
+
+    macro_rules! def_id {
+        () => {
+            if let Some(def_id) = dep_node.extract_def_id(tcx) {
+                def_id
+            } else {
+                // return from the whole function
+                return false
+            }
+        }
+    };
+
+    macro_rules! krate {
+        () => { (def_id!()).krate }
+    };
+
+    macro_rules! force {
+        ($query:ident, $key:expr) => {
+            {
+                use $crate::util::common::{ProfileQueriesMsg, profq_msg};
+
+                // FIXME(eddyb) Get more valid Span's on queries.
+                // def_span guard is necessary to prevent a recursive loop,
+                // default_span calls def_span query internally.
+                let span = if stringify!($query) != "def_span" {
+                    $key.default_span(tcx)
+                } else {
+                    ::syntax_pos::DUMMY_SP
+                };
+
+                profq_msg!(tcx,
+                    ProfileQueriesMsg::QueryBegin(
+                        span.clone(),
+                        ::ty::maps::QueryMsg::$query(profq_key!(tcx, $key))
+                    )
+                );
+
+                match ::ty::maps::queries::$query::force(tcx, $key, span, *dep_node) {
+                    Ok(_) => {},
+                    Err(e) => {
+                        tcx.report_cycle(e).emit();
+                    }
+                }
+            }
+        }
+    };
+
+    match dep_node.kind {
+        // These are inputs that are expected to be pre-allocated and that
+        // should therefore always be red or green already
+        DepKind::AllLocalTraitImpls |
```

Comment:
This match feels mildly unfortunate -- i.e., it *feels* like we ought to be able to derive this information from other macros and not require it to be "re-entered" here. But oh well, I haven't thought about how to actually *do* that. Might be nice to file a FIXME issue about it though.

Reply:
Yeah, it's definitely not pretty. But I didn't want spend too much time on making this macro generated before we have a proof of concept implementation of red/green. I'll add a FIXME.